### ISSUE TITLE
LoopingCall.withCount with interval 0

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8125]
|<img alt="gkonst's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| gkonst reported|
|-|-|
|Trac ID|trac#8125|
|Type|defect|
|Created|2015-11-24 15:59:55Z|
|Branch|https://github.com/twisted/twisted/tree/loopingcall-count-interval-zero-8125|

When interval in LoopingCall.withCount is set to 0 (at start or during execution) exceptions.ZeroDivisionError is thrown in twisted/internet/task.py, line 141, _intervalOf function called by counter decorator.

Setting interval to 0 is acceptable and in such situation countCallable should be called with 1 assuming this is maximum we can get from our loop.

Attachments:

* [loopingcall-with-count-interval-zero-test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a75/a755f5df59f10e12ff8dfef534d48d8d27b72286/e2fe46c697e9720a5b0392b67d2fc629069a1d58.patch) (1041 bytes) - added by gkonst on 2015-11-24 16:02:00Z - test to check interval 0
* [loopingcall-with-count-interval-zero.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a75/a755f5df59f10e12ff8dfef534d48d8d27b72286/db4c6c7cf5691a1c841633578fe8851d3fec01f7.patch) (493 bytes) - added by gkonst on 2015-11-24 16:02:59Z - patched counter function in LoopingCall
* [loopingcall-count-interval-zero-8125.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a75/a755f5df59f10e12ff8dfef534d48d8d27b72286/029ba1586717eaa6d010aea26d4ebb9aaff81723.patch) (4261 bytes) - added by gkonst on 2015-12-02 22:45:38Z - 
* [loopingcall-count-interval-zero-8125v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a75/a755f5df59f10e12ff8dfef534d48d8d27b72286/95b7f5595933043d7e1230d019398f293a7d053e.patch) (2652 bytes) - added by gkonst on 2015-12-04 09:35:58Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8125 8125
type__defect defect
reporter__gkonst gkonst
priority__normal normal
milestone__None None
branch__branches_loopingcall_count_interval_zero_8125 branches/loopingcall-count-interval-zero-8125
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1448380795750681 1448380795750681
changetime__1450010593731985 1450010593731985
version__None None
owner__adiroiban adiroiban
cc__g.k@...
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="gkonst's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|gkonst commented|
|-|-|

Simple patch attached

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

(In [46337]) Branching to loopingcall-count-interval-zero-8125.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to gkonst|
|-|-|

Many thanks for the bug report and the patch. Feel free to submit a single patch contain both the code and the test.

I am not sure if this is the right fix.

I have updated your tests to use a Clock reactor so that we can control the time.

I have also wrote a tests in which the loop call is delayed with 2 seconds. What should happen in this case?

From my understand the countCallable should be called with the number of skipped calls.

How do we handle this case?

Maybe we should update the docstring to inform how the zero value is handled.

Please check my comments and changes and submit a new path on top of the latest branch.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="gkonst's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|gkonst removed owner|
|-|-|

Thanks for quick review.

I'm submitting new version with removed lastTime update if interval is zero.
I don't think we can calculate skipped counts in such case. If you are setting interval to zero you want to go as fast as it can go.

In my case I was using twisted to write some load generator for performance tests. When interval was set to some very small value there was no change in number of transactions being sent. When interval was zero I could sent 3 times more transactions - looks like callLater overhead for scheduling next call was more than preparation of next transaction. 

Idea was to use withCount to check how many calls were skipped and then change interval to zero to let missing calls be done and return to original interval value.

I've updated second test case to show general idea for this - however handling of this should be left to implementation outside library. Also I'm not sure if this is good test case to be left in code.

I updated docstring to include this information.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to gkonst|
|-|-|

> I've updated second test case to show general idea for this - however handling of this should be left to implementation outside library. Also I'm not sure if this is good test case to be left in code.

Is better to move and convert this tests into an example... a http://sscce.org/ would be best as it could help other developers understand your use case and if a refactoring is later required people can understand why we have this special use case for interval=0 

For the test part, is enough if we have a test which checks that callback is still called with 1 even if the clock was advanced.

I have updated a bit the docs, to what I hope, is an easier to read format :)

I have sent the patch to the tests.

Please check my comment and simplify the second test.

Creating a sscce example is optional.

Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="gkonst's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|gkonst removed owner|
|-|-|

I reverted second test to original with modification that all calls should be with 1.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set owner to gkonst|
|-|-|

The patch needs a news file fragment so that this change is advertised in the release notes.

https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

I have created such a file. Please check it and let me know if it correctly described what was done in this patch.

Sorry for this late comment.

--------

What is a 'bi delayed called' ?

----------

Please check my latest commit and let me know if all is OK.

If changes are required please submit a new patch.

As soon as I got your approval I will merge it.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="gkonst's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|gkonst set owner to @adiroiban|
|-|-|

Commit looks OK. I can't find news file.

Thanks a lot.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set owner to gkonst|
|-|-|

Ups. I forgot to push my local change.

Please check now the latest commit.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="gkonst's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|gkonst set owner to @adiroiban|
|-|-|

Looks great!

Thanks for your support.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban set owner to gkonst|
|-|-|

Hi,

Sorry for coming back on this.
Before the merge I did one more review and I found a problem with the case when interval is updated at runtime.

I wrote a test and then a fix.

Please check that all is OK and that you understand the test docstring and its content.

Feel free to send a patch on top of this branch.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="gkonst's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|gkonst set owner to @adiroiban|
|-|-|

Hi,

Sorry for late reply.

Last solution is fine for me too. If you will look at first patch it is exactly the same.

After your comment about counting skipped calls for interval 0 I started to think how this could be done.

Interval 0 is special case (seems like nobody set this before or just didn't raised a ticket for division by zero). My second commit was to do only call with 1 and not update anything else.

So then instead of having skipped count, when you will change to interval != 0 you will get number of calls which would have been done during time when interval was 0. Additionally couple of nanoseconds are saved because we don't have to store _realLastTime.

On the other hand it is easy to define own counter and make new class with it.

Please decide which approach would be better for everyone.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46471]) Merge loopingcall-count-interval-zero-8125: Handle LoopingCall.withCount with interval 0.

Author: gkonst
Reviewer: adiroiban
Fixes: [#8125](https://github.com/twisted/twisted/issues/8125)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban commented|
|-|-|

Sorry for the delay.

I have merged the current revision.

If you think that it contains any regressions please re-open this ticket.

It you want to see more features or found more bugs please open new tickets.

I have no idea why approach is better for everyone :)

Many thanks!
